src/lambda/commands/uploadLambda.ts

```
@@ -0,0 +1,246 @@
+    functionNode: LambdaFunctionNode,
+    window = Window.vscode()
+): Promise<telemetry.Result> {
+    const workspaceFolders = vscode.workspace.workspaceFolders || []
+
+    const parentDirArr = await window.showOpenDialog({
+        canSelectFolders: true,
+        canSelectFiles: false,
+        canSelectMany: false,
+        defaultUri: workspaceFolders[0]?.uri,
+    })
+
+    if (!parentDirArr || parentDirArr.length !== 1) {
+        return 'Cancelled'
+    }
+
+    return await zipAndUploadDirectory(functionNode, parentDirArr[0].fsPath)
+}
+
+async function runUploadLambdaZipFile(
+    functionNode: LambdaFunctionNode,
+    window = Window.vscode()
+): Promise<telemetry.Result> {
+    const workspaceFolders = vscode.workspace.workspaceFolders || []
+
+    const zipFileArr = await window.showOpenDialog({
+        canSelectFolders: false,
+        canSelectFiles: true,
+        canSelectMany: false,
+        defaultUri: workspaceFolders[0]?.uri,
+        filters: {
+            'ZIP archive': ['zip'],
+        },
+    })
+
+    if (!zipFileArr || zipFileArr.length !== 1) {
+        return 'Cancelled'
+    }
+
+    const zipFile = fs.readFileSync(zipFileArr[0].fsPath)
+
+    return await uploadZipBuffer(functionNode, zipFile)
+}
+
+async function runUploadLambdaWithSamBuild(
+    functionNode: LambdaFunctionNode,
+    window = Window.vscode()
+): Promise<telemetry.Result> {
+    const workspaceFolders = vscode.workspace.workspaceFolders || []
+
+    const parentDirArr = await window.showOpenDialog({
+        canSelectFolders: true,
+        canSelectFiles: false,
+        canSelectMany: false,
+        defaultUri: workspaceFolders[0]?.uri,
+    })
+
+    if (!parentDirArr || parentDirArr.length !== 1) {
+        return 'Cancelled'
+    }
+
+    // TODO: Detect if handler is present and error out prematurely?
+    // TODO: What should we do if no manifest (package.json, requirements.txt, etc.) is present? Directory won't be `sam build`-able.
```

anything we can reasonably detect, we should detect. And failing early is the best thing to do, until we have a better approach. Don't need to figure better approach right now unless it's blocking a common case.

---

I think I wrote this TODO before creating the no-build case (I originally ran into complications here when I tried building an imported Lambda that didn't have a package.json). I think I'll remove the comment, have no action taken here, and allow this to fail at the `sam build` step below.